jbmc/src/java_bytecode/java_trace_validation.cpp

```diff
@@ -0,0 +1,188 @@
+  if(!can_cast_expr<member_exprt>(member_expr))
+    return false;
+  if(!member_expr.has_operands())
+    return false;
+  const auto symbol_operand = get_inner_symbol_expr(member_expr);
+  return symbol_operand.has_value();
+}
+
+static void raise_error(const std::string &side, const exprt &expr)
+{
+  throw std::runtime_error(
+    "JBMC Check trace assumption failure on " + side + " expression:\n" +
+    expr.pretty());
+}
+
+bool valid_lhs_expr_high_level(const exprt &lhs)
+{
+  return can_cast_expr<member_exprt>(lhs) || can_cast_expr<symbol_exprt>(lhs) ||
+         can_cast_expr<index_exprt>(lhs) ||
+         lhs.id() == ID_byte_extract_little_endian;
+}
+
+bool valid_rhs_expr_high_level(const exprt &rhs)
+{
+  return can_cast_expr<struct_exprt>(rhs) || can_cast_expr<array_exprt>(rhs) ||
+         can_cast_expr<constant_exprt>(rhs) ||
+         can_cast_expr<address_of_exprt>(rhs) ||
+         can_cast_expr<symbol_exprt>(rhs) ||
+         can_cast_expr<array_list_exprt>(rhs);
+}
+
+static void check_lhs_assumptions(const exprt &lhs, const namespacet &ns)
+{
+  if(!valid_lhs_expr_high_level(lhs))
+    raise_error("LHS", lhs);
+  // check member lhs structure
+  else if(const auto member = expr_try_dynamic_cast<member_exprt>(lhs))
+  {
+    if(!check_member_structure(*member))
+      raise_error("LHS", lhs);
+  }
+  // check symbol lhs structure
+  else if(const auto symbol = expr_try_dynamic_cast<symbol_exprt>(lhs))
+  {
+    if(!check_symbol_structure(*symbol))
+      raise_error("LHS", lhs);
+  }
+  // check index lhs structure
+  else if(const auto index = expr_try_dynamic_cast<index_exprt>(lhs))
+  {
+    if(index->operands().size() != 2)
+      raise_error("LHS", lhs);
+    if(!check_symbol_structure(index->op0()))
+      raise_error("LHS", lhs);
+    if(!expr_try_dynamic_cast<constant_exprt>(index->op1()))
+      raise_error("LHS", lhs);
+  }
+  // check byte extract lhs structure
+  else if(lhs.id() == ID_byte_extract_little_endian)
+  {
+    if(!check_symbol_structure(lhs.op0()))
+      raise_error("LHS", lhs);
+    if(!expr_try_dynamic_cast<constant_exprt>(simplify_expr(lhs.op1(), ns)))
+      raise_error("LHS", lhs);
+  }
+  else
+  {
+    raise_error("LHS", lhs);
+  }
+}
+
+static void check_rhs_assumptions(const exprt &rhs, const namespacet &ns)
+{
+  if(!valid_rhs_expr_high_level(rhs))
+    raise_error("RHS", rhs);
+  // check address_of rhs structure (String only)
+  else if(const auto address = expr_try_dynamic_cast<address_of_exprt>(rhs))
+  {
+    const auto symbol_expr = get_inner_symbol_expr(*address);
+    if(!symbol_expr)
+      raise_error("RHS", rhs);
+  }
+  // check symbol rhs structure (String only)
+  else if(const auto symbol_expr = expr_try_dynamic_cast<symbol_exprt>(rhs))
+  {
+    if(!check_symbol_structure(*symbol_expr))
+      raise_error("RHS", rhs);
+  }
+  // check struct rhs structure
+  else if(const auto struct_expr = expr_try_dynamic_cast<struct_exprt>(rhs))
+  {
+    if(!struct_expr->has_operands())
```

Comment:
:racehorse:

Reply:
? This isn't to do with endian-ness, this is about verifying structs-within-structs. As it stands if I had `class A { int x; float y; }` then you would check that x and y are constant, but if I had `class B extends A { }` then it would not be verified